File: exchange/RateLimiter.ts
export default class RateLimiter {
  window: [number, number][];
  maxWeight: number;
  constructor(maxWeightPerMin: number = 5000) {
    // 5000, not 6000 — safety margin
    this.window = []; // (timestamp, weight) pairs
    this.maxWeight = maxWeightPerMin;
  }
  canRequest(weight = 1): boolean {
    const now = Date.now() / 1000;
    // Remove entries older than 60s
    while (this.window.length && this.window[0][0] < now - 60) {
      this.window.shift();
    }
    const currentWeight = this.window.reduce(
      (sum, pair) => (sum += pair[1]),
      0,
    );
    return currentWeight + weight <= this.maxWeight;
  }
  record(weight = 1) {
    this.window.push([Date.now() / 1000, weight]);
  }
}

File: exchange/rate_limiter.py
# exchange/rate_limiter.py

import time
from typing import List, Tuple


class RateLimiter:
    """
    Simple sliding-window rate limiter for exchange API requests.
    """

    def __init__(self, max_weight_per_min: int = 5000):
        # 5000, not 6000 — safety margin
        self.window: List[Tuple[float, int]] = []  # list of (timestamp, weight)
        self.max_weight: int = max_weight_per_min

    def can_request(self, weight: int = 1) -> bool:
        now = time.time()
        # Remove entries older than 60 seconds
        self.window = [(ts, w) for ts, w in self.window if ts >= now - 60]

        current_weight = sum(w for ts, w in self.window)
        return current_weight + weight <= self.max_weight

    def record(self, weight: int = 1):
        self.window.append((time.time(), weight))
